### ISSUE TITLE
Prototype self-hosted app bootstraping in 2.x

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3869]
Found myself researching/prototyping this to get further on https://github.com/Starcounter/Starcounter.Core/issues/120. 

We've taken a principal decision to try realizing this in a future 2.x-based version, initially with a limited scope though (basically, one self-hosted app / database), possibly extending that over time.

There are oldish but still relevant background discussions in the .Core tracker in https://github.com/Starcounter/Starcounter.Core/issues/41 and https://github.com/Starcounter/Starcounter.Core/issues/59, and some documentation at [Starcounter.Core wiki, Self-hosted apps](https://github.com/Starcounter/Starcounter.Core/wiki/Apps-on-Starcounter.Core#self-hosted-apps). Also relevant: https://github.com/Starcounter/CompanyTrack/issues/232.

**Comment by per-samuelsson** [comment]
Just for the record: I'm doing this in **pnext** only.

**Comment by per-samuelsson** [comment]
If we decide to actually realize this, we must decide how to front it on the highest level. Two dedicated templates, e.g.:

* Starcounter Application
* Starcounter Self-hosted Application

Or via configuration on the former? Or something else?

Just from the top of my mind, it seem a single customizable template / application would be most reasonable, since how to *run*/*host* an application is more of a management thing, and it make sense to be able to run both modes using a single code base.

**Comment by per-samuelsson** [comment]
Adding to the above: how to handle references to Starcounter assemblies.

In our standard 2.x context, references are resolved without requiring "Copy Local" and we expect that Starcounter assemblies are not part of the application distribution. Resolving them runtime is implicit: since the host is `sccode`, they are all available.

For this model to work, we need to explicitly resolve them at runtime (which is certainly doable).

**Comment by per-samuelsson** [comment]
> Adding to the above: how to handle references to Starcounter assemblies.

I think the way to go here, considering that we likely want to preserve the model for our standard/legacy apps (where binaries are not copied), is that we introduce a custom assembly, e.g. `Starcounter.SelfHost` (or `.Standalone` or whatever we want to call it) containing the front API to bootstrap apps in independent hosts. This assembly is just thin and standard copied to output, and include a resolver that knows how to find assemblies from the installation.

**Comment by per-samuelsson** [comment]
> I think the way to go here [...] is that we introduce a custom assembly, [...] and include a resolver that knows how to find assemblies from the installation.

Bah, just realized that still won't work, since failure to resolve will happen earlier, before we get a chance to install the resolver.

Maybe it's better to review the model we have today actually, and support local copying of our fronting assemblies after all. It's probably involve some work, but is good since it's more in line what can be expected from other frameworks (I think).

**Comment by per-samuelsson** [comment]
> Maybe it's better to review the model we have today actually, and support local copying of our fronting assemblies after all. It's probably involve some work, but is good since it's more in line what can be expected from other frameworks (I think).

Did some reviewing of how projects in the standard framework support transitive dependencies and it seem to work to our advantage. That is, if we introduce a copy-local dependency to a new self-hosting specific API assembly, assemblies referenced from that seem to be copied as desired. 

If that hold true, it gives us a way to configure this new independent hosting style using copy local and still preserve the way we use today for our traditional model.

**Comment by per-samuelsson** [comment]
Regarding the weaver strategy we have for this (https://github.com/Starcounter/Starcounter.Core/issues/120), one thing that involve extra work is if we want to support unit testing of standalone Starcounter *libraries* (i.e. not only exe assemblies). Which I guess we must?

**MyLib.dll**
```cs
using Starcounter;

[Database] public class Foo {
  public bool Approve() { ... } 
}
```

**MyLib.Tests.dll**
```cs
using MyLib;
using xUnit;

class TestSetup {
  public TestSetup() {
    // Somewhat pseudo-coding
    var host = new AppHostBuilder().UseDatabase("default").Build();
    host.Run();
  }
}

public class FooTests : ITestFixture<TestSetup> {
  [Fact]
  public void TestApprove() {
    Db.Transact(() => {
      var f = new Foo();
      Assert.True(f.Approve());
    });
  } 
}
```

To support that, we must redesign some core parts of how the weaver operate. First of all, it accept only executables today. That should be quite easy to change though. Second, it operates on **fully resolved applications**, i.e. you pass in the main assembly, and the weaver will weave and resolve that and *everything it reference*, assuming nothing has been weaved before. That model will have to change to support the above, since the library could end up as a reference to another library or exe that is passed to the weaver, and in that case, the first library will already have been pre-weaved (compile-time).

**Comment by per-samuelsson** [comment]
Kind of struggle a bit with the question I raised in this earlier comment: https://github.com/Starcounter/Starcounter/issues/3869#issuecomment-252209965. Whether to support self-hosted applications via a configuration in a normal "Starcounter Application" project. Or if it should be a new project type of its own.

First I felt the former was the correct choice.

Now I'm leaning towards that a dedicated project type is favorable:

1. We need to copy dependencies locally. If we have just one project type, we need to start doing that for "standard" applications too.
2. With a separate project type, we don't need to risk messing up for standard projects. This is good since the upcoming changes will be quite major under the surface.
3. Switching back and forth between "standard" and "self-host" via some property in a single project type will need to fiddle quite a lot with the build pipeline and already built artifacts (and reference resolvning). With a custom project type, we don't need this.
4. It will still be possible to support both runtime modes for a single application. Just move the bulk of the code to a shared library, and reference it from both "bootstrap projects".
5. Self-hosting apps will need a self-bootstrap sequence, while standard projects do not. It make more sense to host that in a separate entrypoint (rather than one that need to "switch" on what the host is).

**Comment by per-samuelsson** [comment]
> It will still be possible to support both runtime modes for a single application. Just move the bulk of the code to a shared library, and reference it from both "bootstrap projects".

Hm, realized it's not that easy actually. Stuff like www-files and `weaver.ignore` is hard to keep in a library. That's clearly on the negative side with a separate project type.

**Comment by per-samuelsson** [comment]
> Now I'm leaning towards that a dedicated project type is favorable

Nah, I'm back again at my first stand: single project type "Starcounter Application". We change so that assemblies are always copied locally (and deal with any problem that might come with) and allow developers to configure (and toogle) different modes of hosting.

**Comment by per-samuelsson** [comment]
Hm, when I've glued compile time weaving and self-hosting together, I'm getting this when bootstrapping:

> FatalExecutionEngineError occurred
  HResult=-2146233082
  Message=Exception of type 'System.ExecutionEngineException' was thrown.
  InnerException: 

> Additional information: The runtime has encountered a fatal error. The address of the error was at 0x0417bbec, on thread 0x3c10. The error code is 0xc0000005. This error may be a bug in the CLR or in the unsafe or non-verifiable portions of user code. Common sources of this bug include user marshaling errors for COM-interop or PInvoke, which may corrupt the stack.

That one doesn't feel particularly great 😎 

It's triggered on `delegateArgTypes.Add(RestDelegateArgumentTypes.REST_ARG_SESSION);` in `\src\Starcounter.Rest\UriManagedHandlersCodegen.cs` but it's not really likely the problem is in that API I guess.

**Comment by per-samuelsson** [comment]
> FatalExecutionEngineError occurred

I've spent a considerable time the last days trying different approaches to track this bug down, but I just can't find what goes wrong. 

The only positive thing is I've narrowed it down to a simple use-case and that it's reproducible 100% in that setup - compile a simple self-hosted application, weave it, and run the weaved executable.

I've found that it only occur if the self-hosted app utilize the gateway, so right now I'm putting this bug aside by forcing such apps not to use the gateway, and I will take a look at it again later on.

**Comment by per-samuelsson** [comment]
Added **Technical dept** section to OP.

**Comment by per-samuelsson** [comment]
Bumped into another problem I didn't see coming: 

When I configure a Starcounter application project to be self hosted, I need to explicitly bootstrap the process for the executable, which is of course not a problem. But attaching the debugger to it is only possible once it has entered the "running" state, and then F11 / Step into doesn't work properly, and neither stepping errors that happen early in the process life, before the debugger has been attached.

I'm pushing this forward as a technical dept for now, since I don't know how to handle it yet.

**Comment by per-samuelsson** [comment]
One concern I have with this new self-hosted app model is that since we need to switch from not copying Starcounter dependencies locally to do just that, and since our bits are not very well separated in all parts, each project output directory will be pretty bloated with Starcounter assemblies.

Right now, I'm requiring any Starcounter Application that want to host itself to have two references: `Starcounter` and `Starcounter.Bootstrap`. (FWIW, I'm planning to trim this down to one). The result of that is that just compiling the app will end up with this:

```
bin\Debug:
Newtonsoft.Json.dll
Newtonsoft.Json.xml
Simplified.Data.Model.dll
Simplified.Data.Model.pdb
Starcounter.dll
Starcounter.Apps.JsonPatch.dll
Starcounter.Bootstrap.dll
Starcounter.Extensions.dll
Starcounter.Extensions.pdb
Starcounter.Hypermedia.dll
Starcounter.Internal.dll
Starcounter.Logging.dll
Starcounter.Server.Rest.dll
Starcounter.XSON.dll
Starcounter.XSON.pdb
StarcounterApplication18.exe
StarcounterApplication18.pdb
```

**Comment by per-samuelsson** [comment]
Expanding on above concern: actually, that transient references get pulled in is not optimal, but at least somewhat justified. It's a lot harder to explain how those "special assemblies" of ours end up in there (because of weaving) and I really think that must be addressed first.

In the shared host, users never really see the consolidated runtime directory (where all weaved artifacts end up). With this new model they will, and we need to adjust to that. I'll see if I can come up with some tweak that allow the self-hosted host to resolve our "special assembly" references from somewhere else instead.

**Comment by per-samuelsson** [comment]
> I'll see if I can come up with some tweak that allow the self-hosted host to resolve our "special assembly" references from somewhere else instead.

Thinking more about that, it seem to me that the most correct thing to do really, is to weave those "libraries with database classes" upfront, probably install-time, and have all flavors of the host resolve the weaved versions from there. That would benefit startup time of applications, and also not bloat the disk.

The problem with that is the deadline we have here, since that would mean some more work.

**Comment by per-samuelsson** [comment]
> Expanding on above concern: actually, that transient references get pulled in is not optimal, but at least somewhat justified. **It's a lot harder to explain how those "special assemblies" of ours end up in there (because of weaving) and I really think that must be addressed first.**

I have come up with a generic redesign that I've implemented that seem to work. What I've done is that I've made the code host assembly resolver to an abstraction and made a custom resolver for self-hosting scenarios. The resolver there will know how to resolve files - assemblies and schema files - not only from the root application directory, but also from any set of other directories too. Hence, we can store "special assemblies" and schemas in the local weaver cache, and have the host consume them transparently from there.

**Comment by per-samuelsson** [comment]
>I have come up with a generic redesign that I've implemented that seem to work. What I've done is that I've made the code host assembly resolver to an abstraction and made a custom resolver for self-hosting scenarios. The resolver there will know how to resolve files - assemblies and schema files - not only from the root application directory, but also from any set of other directories too.

New setback - that worked nice and smooth when just running that self-hosted application. But once I set the same thing up with different unit testing frameworks, it didn't work at all. After I researched, it turns out these frameworks (both NUnit and xUnit) copies the assembly to some other directory before testing, and hence they need to be fully self-contained to work. I need a new fresh approach on this. Seem our model with serialized schema files will be hard to get working in combination with unit tests.

**Comment by per-samuelsson** [comment]
> After I researched, it turns out these frameworks (both NUnit and xUnit) copies the assembly to some other directory before testing, and hence they need to be fully self-contained to work. I need a new fresh approach on this. Seem our model with serialized schema files will be hard to get working in combination with unit tests.

#3941

**Comment by per-samuelsson** [comment]
As for

>  Attach debugger when bootstrapping self-hosted application (temporary disabled in 1668656fda75a392c057d7c5e29859583f4e1be8)

from the OP technical dept list, I asked @linkdata for advice and he suggested P/Invoke over Windows `CreateProcess` to create the process with the main thread in suspended state.

After researching that track a bit, that seem to be the suggested way to do it. I'll try to implement that ASAP. If it works, it's a relief since it was the main thing I was worried about and had no real clue to get further on.

**Comment by per-samuelsson** [comment]
> After researching that track a bit, that seem to be the suggested way to do it. I'll try to implement that ASAP. If it works, it's a relief since it was the main thing I was worried about and had no real clue to get further on.

Worked like a charm. Now it's possible to debug self-hosted apps in the same straightforward manner as you would do any standard console application. Thanks for the hint, @linkdata. 👍

**Comment by per-samuelsson** [comment]
New debt listed in OP, one I'll need help to figure out:

> Decide what should be the default references in Starcounter Application and Library project templates.

**Comment by per-samuelsson** [comment]
Request for feedback on unit testing story and hosting API: https://github.com/Starcounter/RebelsLounge/issues/54.

**Comment by per-samuelsson** [comment]
One small step forward: first time succeeded doing an extended daily build. Next up is nighlies.

**Comment by per-samuelsson** [comment]
After deciding #3984, I'm altering how to specify if an app should be self-hosted or bootstrapped into the shared host - from being a project-based configuration switch to a command-line option, possible to specify on `Debug | Command line arguments`.

To be consistent, I'm adding also the ability to launch a self-hosted application via `star.exe`, using a unified option. I guess the normal workflow would be to launch the exe though.

**Comment by per-samuelsson** [comment]
> To be consistent, I'm adding also the ability to launch a self-hosted application via star.exe, using a unified option. I guess the normal workflow would be to launch the exe though.

I decided to postpone this, or alternatively, not support it at all. At least not until we find it necessary. Bootstrapping a self-hosted app from the command-line is restricted to launching the exe straight from the OS.

**Comment by per-samuelsson** [comment]
Boosting the friday feeling with another milestone: after like 10 attempts, I've been able to successfully run a nightly build (after adapting the build system, but not disabled any tests).

That would mean this thing is finally good to merge -mainly waiting some PR's on sample/prefab repositories that I have adapted and that was not approved yet.

**Comment by per-samuelsson** [comment]
Finally merged, along with new build-system branch and dependent repos: https://github.com/Starcounter/Starcounter/commit/d5aa1ff89fbffba7842275e2e3132c4b19b01ea6.

Going into this now:

* Announce stuff
* Write documentation on self-hosting, unit testing
* Go through and close issues
* Delete remote branches
* Write issues for technical debts

Will close this down after that.
